docs/what-is-wrapped.md

```diff
@@ -0,0 +1,1014 @@
+    if (!Utils.isValidEmail(user.email)) throw new RuntimeException("email is invalid")
+    repo.save(user)
+    val message = if (laws.canHaveRussianPassport(user.age)) "hello big buddy, you are saved" else "hello, you are saved"
+    emailService.sendEmail(user.email, message)
+  }
+}
+```
+The problem here is that we have to make a lot of duplicate checks. `UserService.saveUser` checks that `email` and `age` are valid, otherwise it can't save it to DB.
+Then in calls `LawsInfoService.canHaveRussianPassport` and `EmailService.sendEmail`, and they check that `email` and `age` are valid as well.
+It's kinda stupid because at that point we already checked that. However, we can't remove these checks because `sendEmail` or `canHaveRussianPassport` can be called anywhere which means we can't assume that input data is always valid.<br>
+Hence, we need some way to validate input data and then "remember" that this data is valid. That's where our `Wrapped` type comes into play. We can make such checks during `Wrapped` creation.
+In this case we need to check input data only once and then guarantee that either we don't have a `Wrapped` object, or it contains a valid data.<br>
+This concept is called `smart constructors`. In short, smart constructors are useful when you cannot guarantee that you can create a valid object based on provided data.
+For example, we can create an `Email` from `String`, but we can't do it for every `String`, because not every `String` is a valid `Email`.
+```scala
+final case class Email(value: String)
+val email: Email = Email("not an email") // this should not be allowed
+```
+We have to somehow validate input data. One way is to throw an exception during the construction.
+```scala
+final case class Email(value: String) {
+  private val dummyEmailRegexp: String = "(\\w|\\d)+@(\\w|\\d)+\\.(\\w|\\d)+"
+  require(value.matches(dummyEmailRegexp), s"email must match regexp $dummyEmailRegexp")
+}
+val email: Email = Email("not an email") // compiles, but throws exception at runtime, i.e. object is not created
+```
+However, we don't throw exceptions in Scala. Instead, we should hide constructor and use factory method that will be the only way to create an instance of our class.
+```scala
+final class Email private (val value: String)
+object Email {
+  private val dummyEmailRegexp: String = "(\\w|\\d)+@(\\w|\\d)+\\.(\\w|\\d)+"
+  def fromString(value: String): Either[String, Email] =
+    if (value.matches(dummyEmailRegexp)) Right(new Email(value)) else Left(s"email must match regexp $dummyEmailRegexp")
+    
+}
+//val email: Email = new Email("not an email") // won't compile anymore
+//val email: Email = Email.fromString("not an email") // won't compile anymore
+val email: Either[String, Email] = Email.fromString("not an email")
+```
+Notice that we don't use `case class` anymore. This is because `case class`es by their nature are open data structures that can be easily created. We don't want it anymore.
```

Comment:
It's not a remark, but an addition. It's possible to make case classes not so open and with removed `copy` method.
For example
```scala
sealed abstract case class Email private (value: String)

object Email {
  def fromString(emailStr: String): Option[Email] = {
    Some(new Email(emailStr) {} )
  }
}
```

And in this case hashcode and equals I assume still exist and work correctly.

Reply:
Thanks for reminder, you are right. I've added your example to document.